src/shell/xdg/popup.rs

```diff
@@ -88,42 +87,53 @@ impl Popup {
             + Dispatch<xdg_popup::XdgPopup, PopupData>
             + 'static,
     {
-        let arc = Arc::new(OnceCell::new());
+        // We really need an Arc::try_new_cyclic function to handle errors during creation.
+        // Emulate that function by creating an Arc containing None in the error case, and use the
+        // closure's context to pass the real error back to the caller so the invalid Arc is never
+        // returned.
+        let mut err = Ok(());
+        let inner = Arc::new_cyclic(|weak| {
+            let surface =
+                XdgShellSurface::new(wm_base, qh, surface, PopupData { inner: weak.clone() })
+                    .map_err(|e| err = Err(e.into()))
+                    .ok()?;
+            let xdg_popup = surface
+                .xdg_surface()
+                .get_popup(parent, position, qh, PopupData { inner: weak.clone() })
+                .map_err(|e| err = Err(e.into()))
+                .ok()?;
 
-        let surface =
-            XdgShellSurface::new(wm_base, qh, surface, PopupData { inner: Arc::downgrade(&arc) })?;
-        let xdg_popup = surface.xdg_surface().get_popup(
-            parent,
-            position,
-            qh,
-            PopupData { inner: Arc::downgrade(&arc) },
-        )?;
-        arc.set(PopupInner {
-            surface,
-            xdg_popup,
-            pending_position: (AtomicI32::new(0), AtomicI32::new(0)),
-            pending_dimensions: (AtomicI32::new(-1), AtomicI32::new(-1)),
-            pending_token: AtomicU32::new(0),
-            configure_state: AtomicU32::new(PopupConfigure::STATE_NEW),
-        })
-        .unwrap();
-        Ok(Popup { inner: arc })
+            Some(PopupInner {
+                surface,
+                xdg_popup,
+                pending_position: (AtomicI32::new(0), AtomicI32::new(0)),
+                pending_dimensions: (AtomicI32::new(-1), AtomicI32::new(-1)),
+                pending_token: AtomicU32::new(0),
+                configure_state: AtomicU32::new(PopupConfigure::STATE_NEW),
+            })
+        });
+        debug_assert!(inner.is_some() || err.is_err());
```

Comment:
Personally I think debug_assert is too weak. If it does return Err, then there is probably state corruption.

Reply:
This assert is something that can never happen (with the current code) - it's mainly there to help document the assumption, and doesn't need to be there in production.

Reply:
I agree, but could you add a comment explaining that then?